### ISSUE TITLE
Time format for Spark to Presto conversion of unix_timestamp not inclusive of microseconds

### ISSUE THREAD

**Issue by ddelzell** [tobymao/sqlglot#3796]
Consider the following Spark code
```
select unix_timestamp(mytime)
from table1
```
This will be transpiled to 
```
SELECT TO_UNIXTIME(COALESCE(TRY(DATE_PARSE(CAST(mytime AS VARCHAR), '%Y-%m-%d %T')), PARSE_DATETIME(CAST(mytime AS VARCHAR), 'yyyy-MM-dd HH:mm:ss'))) 
FROM table1
```

However, if `mytime` contains milliseconds, an "INVALID_FUNCTION_ARGUMENT: Invalid format" exception will be thrown.  Adding milliseconds to the casting ('yyyy-MM-dd HH:mm:ss.SSS') solves the issue.

**Comment by tobymao** [comment]
@ddelzell PR appreciated

**Comment by JtoTheGoGo** [comment]
I did some background digging into the source code for this translation and think I found the root cause.
(sorry if I'm mistaken.)

This appears to occur due to the [TIME_FORMAT](https://github.com/tobymao/sqlglot/blob/44d650637d5d7a662b57ec1d8ca74dffe0f7ad73/sqlglot/dialects/hive.py#L234) being applied during the Spark -> Presto translation (which utilizes Hive time formatting rules for **StrToUnix**)

The [StrToUnix](https://github.com/tobymao/sqlglot/blob/44d650637d5d7a662b57ec1d8ca74dffe0f7ad73/sqlglot/expressions.py#L5842C7-L5842C16) method sql generator callback for Presto  [strtounix_sql](https://github.com/tobymao/sqlglot/blob/44d650637d5d7a662b57ec1d8ca74dffe0f7ad73/sqlglot/dialects/presto.py#L584) is where the formatting rules appear to be passed in.

**Comment by VaggelisD** [comment]
Hey @JtoTheGoGo, thanks for looking into this. I've been researching this since yesterday and have come to these conclusions:

The `Hive.TIME_FORMAT` is the default time format as indicated [here](https://spark.apache.org/docs/latest/api/sql/#unix_timestamp) for example. Changing it so that it includes milliseconds will not only alter the dialect semantics but will then break timestamps without milliseconds for instance.
 
I'm assuming that `mytime` in this case is not a STRING column (i.e it's a DATE/TIMESTAMP) in which case Spark ignores the format, otherwise it would not execute similarly to Presto/Trino e.g:

- Cannot execute because there's no ms in the default format `"yyyy-MM-dd HH:mm:ss"`
```
SELECT UNIX_TIMESTAMP('2024-07-24 11:39:42.155')
```

- Executes fine because the user-inputted format is matched
```
SELECT UNIX_TIMESTAMP('2024-07-24 11:39:42.155', 'yyyy-MM-dd HH:mm:ss.SSS')
```

- Also executes fine because the input is a TIMESTAMP, so the format is ignored
```
SELECT UNIX_TIMESTAMP(CAST('2024-07-24 11:39:42.155' AS TIMESTAMP), 'yyyy-MM-dd HH:mm:ss.SSS')
```


So to summarize, the problems are: 
1. In Spark, `UNIX_TIMESTAMP(x, fmt)` will execute fine if `x` is a TIMESTAMP and it will ignore `fmt`
2. Coming to Presto/Trino, we don't know the type of `x` but we presume `fmt` is it's proper/matching format; We attempt to reparse `x` into a TIMESTAMP according to Presto/Trino's version of `fmt` before passing it through `TO_UNIXTIME()`, but in this case it won't match.

If we can find a standard way to bridge (1) -> (2) we'll be able to solve this across the board. Another solution would be to include your format in `UNIX_TIMESTAMP()` which will get transpiled properly to Presto/Trino and the query will execute afaict.

**Comment by VaggelisD** [comment]
I may have stumbled upon a patch for this which is applying `DATE_FORMAT` to trim off the milliseconds & timezone from the timestamp (shouldn't offset the unix timestamp by more than +-1 second afaict), so the transpilation looks like this:

```
SELECT TO_UNIXTIME(COALESCE(TRY(DATE_PARSE(CAST(mytime AS VARCHAR), '%Y-%m-%d %T')), PARSE_DATETIME(DATE_FORMAT(mytime, '%Y-%m-%d %T'), 'yyyy-MM-dd HH:mm:ss')))
```
 
If I replace `mytime` with `CURRENT_TIMESTAMP` this patched transpilation works, which is not the case for the initial one.  Can you please verify if that would work on your column?

**Comment by ddelzell** [comment]
That seems to work, except I'm concerned about the stripping of the time zone.  The Spark [function](https://spark.apache.org/docs/3.1.3/api/python/reference/api/pyspark.sql.functions.unix_timestamp.html) utilizes timezone, but `sqlglot` wouldn't have any way of knowing what that is, anyway, right?  So it's a moot point?

**Comment by VaggelisD** [comment]
Generally, in cases where a function `FOO(bar)` accepts a parameter `bar` with uncertain type/value it'll be difficult to transpile properly, as almost always it requires type annotations and/or knowledge of what the underlying value is. 

Regarding the timezone, I think it's OK. More specifically:

- [Spark's](https://spark.apache.org/docs/latest/sql-ref-datatypes.html) `TIMESTAMP` has local timezone:

_TimestampType: Timestamp with local time zone(TIMESTAMP_LTZ). It represents values comprising values of fields year, month, day, hour, minute, and second, with the session local time-zone_

- [Presto's](https://prestodb.io/docs/current/language/types.html#timestamp) / [Trino's](https://trino.io/docs/current/language/types.html#timestamp) `TIMESTAMP` also renders time in the local timezone:

_Instant in time that includes the date and time of day without a time zone. Values of this type are parsed and rendered in the session time zone._

If the Spark and Presto queries are ran in the same region / both sessions set a common TZ, I believe they'd produce similar results.

**Comment by ddelzell** [comment]
That sounds very reasonable, I would support the change to `DATE_FORMAT` to replace the second `CAST` statement.
